DATAES630 - Remove GetResultMapper and friends from core package.

sothawo opened the pull request:
This is a draft PR not ready to be merged.

@mp911de I have moved the ResultMapper stuff into `MappingElasticsearchConverter` and cleaned up some stuff.
Can you check if this what you thought about the rewriting of the mapping? There still is quite some cleanup to be done, refactoring, poilshing etc, but I'd like to have your feedback before going on.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/ResultsMapper.java:
```diff
@@ -36,11 +35,10 @@
  * @author Mark Paluch
+ * @author Peter-Josef Meisch
  */
 public interface ResultsMapper extends SearchResultMapper, GetResultMapper, MultiGetResultMapper {
```
Did this. There are tests in `ElasticsearchTemplateTests` which rely on custom result mappers, especially checking the highlight result or request. I uncommented these so the builds don't fail, but added todos to them, because they have to be checked.

sothawo commented:
I now cleaned up the ResultMappers, the next thing will be cleaning up the whole bunch of map methods into the `MappingElasticsearchConverter`.
Lets see how much more we want to make in this ticket, because I'd like to get this restructuring of the interfaces into `MappingElasticsearchConverter` finished pretty soon to prevent merge conflicts or heavy merging with other PRs that work with the *Template classes. We should perhaps for the other findings create follow-up tickets.

sothawo commented:
Another thought about returning the additional info for search results like the score - there is also the highlight info -: Couldn't we introduce a `SearchResult<Entity>` class that wraps the currently returned entity and additionally has fields for score, highlight etc? Like the `SearchDocument`? The methods of `ElasticsearchOperations` then would return these instead of the entity itself. But how can this be handled in repositories? Should they thenas well reurn these objects?

mp911de commented:
No need to tackle the annotation-based aspects here, it was something that I noticed during the review. We can keep the implementations of the mappers for now in the converter, but we should address these.

Regarding the search results, we have a similar setup with [`GeoResults`](https://github.com/spring-projects/spring-data-mongodb/blob/8cc0609190de1e09aead9ee73bac990ab13bcb11/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoOperations.java#L670) in other places. `GeoResults` is basically an `Iterable<GeoResult<T>>` where auxiliary properties are held that are returned as part of the query. I didn't think about such an arrangement before and now that you've mentioned it, it makes absolute sense to introduce `SearchResults<T>` that is `Iterable<SearchResult<T>>`.

For repositories, we can be more lenient about the API: Users can either declare a query method returning `List<T>` or `SearchResults<T>` if they require additional search-related details.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java:
```diff
@@ -5,21 +5,30 @@
 	private static final Logger LOGGER = LoggerFactory.getLogger(AbstractElasticsearchTemplate.class);
 
-	protected ElasticsearchConverter elasticsearchConverter;
+	protected MappingElasticsearchConverter elasticsearchConverter;
```
done.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java:
```diff
@@ -5,21 +5,30 @@
 
-	protected String buildMapping(Class<?> clazz) {
+    protected MappingElasticsearchConverter createElasticsearchConverter() {
+        MappingElasticsearchConverter mappingElasticsearchConverter = new MappingElasticsearchConverter(new SimpleElasticsearchMappingContext());
```
keep the reference in the `AbstractElasticsearchTemplate`?

sothawo replied:
In the `*Template` classes we can laways get the `MappingContext` from the `ElasticsearchConverter`if needed

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplate.java:
```diff
@@ -1440,7 +1410,7 @@ private IndexRequest prepareIndex(IndexQuery query) {
 					indexRequest = new IndexRequest(indexName, type);
 				}
-				indexRequest.source(resultsMapper.getEntityMapper().mapToString(query.getObject()),
+				indexRequest.source(elasticsearchConverter.mapToString(query.getObject()),
```
removed it.

sothawo commented:
I removed all the elasticsearch stuff from the `MappingElasticsearchConverter`, this is handled in the DocumentAdapters now (which I moved to a separate package). We no have the `Document` stuff dealing with ES, and `Converter` only deals with converting between entity and Document.

In the Converter I renamed the `map...`methods to have the target type in the name like `mapToDocument` or `mapToEntity`. I find that clearer especially when for example mapping a `Document`to an entity or a `SearchDocument`to a `SearchResult`- but might conflict with naming in Converters in other spring data modules.

I added a first small test implementation of `SearchResult<T>` and `SearchResults<T>` along with a `fooSearch` method in `ElasticsearchOperations` and in the template implementations so you might have a look what I think this could be.

We could move all the stuff like fields that are not in the entity (scripted), score, highlight and whatever else is coming from Elasticsearch into the `SearchDocument` and from there into the `SearchResult<T>` enabling the client to have all this information.
Introducing this will change almost all methods in the `ElasticsearchOperations` interface, but I think that will be the only way to get all this additional information back to the user.

Two questions about the `SearchResults<T>`which I defined as `Iterable<SearchResult<T>>`:
1. how do I create a concrete List that can be cast to this?
2. won't returning this conflict with the method of `Iterable<T> search(QueryBuilder query);` in `ElasticsearchRepository`?

sothawo commented:
One more thought: I introduced `SearchDocumentResponse` as to wrap the `SearchDocument`s alomg with information about the search itself (totalHits, maxScore etc). Might be worth a thought to put something along the `SearchHits` to be able to return this as well.

mp911de commented:
> In the Converter I renamed the `map...`methods to have the target type in the name like `mapToDocument` or `mapToEntity`. I find that clearer especially when for example mapping a `Document`to an entity or a `SearchDocument`to a `SearchResult`- but might conflict with naming in Converters in other spring data modules.

Spring Data Commons ships with `EntityReader` and `EntityWriter` interfaces to reflect the direction how entities are read and written. Let's see that we can use these at some point in time.

Re: `SearchResult`: Let's move that effort in the context of a separate ticket.

> won't returning this conflict with the method of `Iterable<T> search(QueryBuilder query);` in ElasticsearchRepository?

We should ask rather whether that method should be exposed on the repository in the first place. 
I'd expect mostly that application code defines methods returning `SearchResult<T>` with the appropriate method signature. The repository implementation needs to check for the return type and return the appropriate result type.

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverter.java:
```diff
@@ -649,7 +674,134 @@ private boolean isSimpleType(Class<?> type) {
+	private <T> void populateScriptFields(T result, SearchHit hit) {
+		if (hit.getFields() != null && !hit.getFields().isEmpty() && result != null) {
+			for (java.lang.reflect.Field field : result.getClass().getDeclaredFields()) {
+				ScriptedField scriptedField = field.getAnnotation(ScriptedField.class);
```
this will be rewritten once we have the `SearchHit<T>` in place, the scripted fields should go in there

sothawo commented:
I removed the tests that were using a custom EntityMapper to extract the highlights; that will be addressed in a different ticket for the `SearchHit<T>` class. I further removed from `ElasticsearchOperations` the methods taking an extra `ElasticsearchConverter` parameter as they are not needed anymore.
So this should be ready for a PR

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/convert/ElasticsearchConverter.java:
```diff
@@ -48,4 +56,47 @@ default String convertId(Object idValue) {
+	 * Get the configured {@link ProjectionFactory}. <br />
+	 * <strong>NOTE</strong> Should be overwritten in implementation to make use of the type cache.
+	 *
+	 * @since 3.2
```
Nit: 4.0
